### ISSUE TITLE
map_read needs optimization

### ISSUE THREAD

**Issue by Ullaakut** [NySwann/NCurses_Sokoban#5]
```
  map = map_create((x = 0) + 1, (y = 0) + 1);
  while ((read_len = read(fd, &buffer, 1)) > 0)
    {
      if (buffer != '\n')
	{
	  if (x >= map->width)
	    map = map_extend_width(map);
	  map->tiles[y][x++] = buffer;
	}
      else
	{
	  if (++y >= map->height)
	    map = map_extend_height(map);
	  x = 0;
	}
    }
```

Since `map_extend_width` and `map_extend_height` are copying the whole map into a new allocated map and destroying the old one at each iteration, this loop is terrible. Especially since it's really easy to fix, you just read the whole file, stock the X and Y of your file and directly create the appropriate map.

**Comment by NySwann** [comment]
Yep, I had to choose between reading the file twice or dynamically resizing the map while reading it.. My lazyness told me to do this ugly thing xD

(Ofcourse i'll improve this terrible code, thanks !)

**Comment by Ullaakut** [comment]
Why not just do 

```
  width = 0;
  height = 0;
  while ((read_len = read(fd, &buffer, 1)) > 0)
    {
      if (buffer != '\n')
	{
	  if (x >= map->width)
	    map = map_extend_width(map);
	  map->tiles[y][x++] = buffer;
	}
      else
	{
	  if (++y >= map->height)
	    map = map_extend_height(map);
          if (!width) {
            width = x;
          } else if (width != x) {
            // warning map chelou
          }
	  x = 0;
	}
    }
  height = y;
  map = map_create(width, height);
```

Maybe I'm missing something though.

**Comment by NySwann** [comment]
The problem is that we need to fill the map with the symbols found in the file, so we need to read the file twice, once for the size of the map, and once to fill the map with correct symbols. To avoid the two open/close required, i did this ..

**Comment by NySwann** [comment]
(That's the goal of this Github repo yeah. Thanks! :D)

**Comment by Ullaakut** [comment]
Why don't you do it in the same loop? On each iteration if there are symbols you store them in a list of found symbols with their position, then you create your empty map of the right size and then populate it with the list of symbols you gathered :-)

**Comment by NySwann** [comment]
Yes, but the list i will be using to store the symbols and their position will need to have a dynamic size because we don't know the number of each symbol before the complete reading of the file (and we can't even stat() or lstat() the file so we're completely fucked). Unless i implement chained lists. So it's not as simple as that. :P

**Comment by Ullaakut** [comment]
Yes, I'd suggest a linked list for the symbols. Since you will not need to do other operations than to append elements to it and then read it at the end, the fact that it's a linked list will not be a performance problem, and it will solve your issue of reading twice / deleting-reallocating your map all the time

**Comment by NySwann** [comment]
Yep, i'll write that soon then. Thanks again :)
